src/test/components/TrackScreenshots.test.js

```diff
@@ -33,7 +33,7 @@ import { getProfileWithNiceTracks } from '../fixtures/profiles/tracks';
 
 // Mock out the getBoundingBox to have a 400 pixel width.
 const TRACK_WIDTH = 400;
-const LEFT = 100;
+const LEFT = 200;
```

Comment:
My theory about why it's needed is that now the screenshot is bigger, and so we need to start a bit further right. I'll check this assumption tomorrow.

Otherwise I did a quick check and this looks very good! It would be nice to have a test testing the new behavior. I can also have a look tomorrow to give some more guidance if you need it, but in the mean time you can look at what the existing tests do and try to add one more by yourself?

Reply:
Okay I'm looking at this. But can you explain what exactly the test should do? Also I noticed something in this part of the code,
```js
it('makes sure the hover image does not go off the top side of screen', () => {
    const { moveMouseAndGetTop } = setup();
    const pageX = LEFT;
    expect(moveMouseAndGetTop(pageX) >= 0).toBe(true);
  });
```
From line 90-94 in this file, isn't `pageX` need to have `TOP` assigned? After I changed this there was an error, but then I increased the value of TOP it passed the test, probably for the bigger screenshots(?)
I actually don't understand the logic behind assigning the values in `LEFT' and `TOP`. I know it's used to test the positioning of the hover, I don't understand how to be sure what constant value is required here. @julienw

Reply:
OK, let me try to explain why we had a bad result with this LEFT=100. And then we can look whether we can keep LEFT=200 or change other constants. (note: the real value in our app is LEFT=150).

So this is the rough layout:

```
TOP of the browser's viewport (aka window)
------------------------------  ^
| HEADER                       TOP
------------------------------  v
|  TITLE  |  SCREENSHOT TRACK
------------------------------
 < LEFT  >
```
In the test, we fix these values in https://github.com/firefox-devtools/profiler/blob/a16fb93e5c6913f812942f9f3af15c8356bc3284/src/test/components/TrackScreenshots.test.js#L35-L37 and https://github.com/firefox-devtools/profiler/blob/a16fb93e5c6913f812942f9f3af15c8356bc3284/src/test/components/TrackScreenshots.test.js#L165-L177:
* distance between the top of the screen track and the top of the browser's viewport is 7
* height of the screenshot track is 50
=> With the previous code, the max height of a screenshot tooltip in this context is `7 + 50/2 = 32px`. With the new code it's: https://github.com/firefox-devtools/profiler/blob/a16fb93e5c6913f812942f9f3af15c8356bc3284/src/components/timeline/TrackScreenshots.js#L159

In tests, the screenshots also have a fixed size: https://github.com/firefox-devtools/profiler/blob/a16fb93e5c6913f812942f9f3af15c8356bc3284/src/test/fixtures/profiles/processed-profile.js#L1034-L1045

So, with the previous code, the size of 150px would be reduced, as was the width to keep the ratio, and this would give: height = 32px (the max) and width = 64px. But with your new code the height of the screenshot will still be 150px, and its width still 300px.

And now we see why we have the issue with `LEFT=100`. We were simulating moving the mouse close to the left border in https://github.com/firefox-devtools/profiler/blob/a16fb93e5c6913f812942f9f3af15c8356bc3284/src/test/components/TrackScreenshots.test.js#L73-L75. In that position, with a width of 300px, the image would be bigger than the available space, so its `left` value would be 0. Same when moving it at LEFT+10 and LEFT+20.

BUT I find it interesting that we test this behavior: that the screenshot tooltip actually stays stuck at the left, instead of having a negative left value! This is what the test called `'makes sure the hover image does not go off the left side of screen'` is supposed to do but I think it's not really good.

So here is my suggestion:
1. Leave LEFT=100,
2. Fix the failing test.You can use "moveMouse" at LEFT, LEFT+10 until LEFT+50, they should all give 0 for result, __but__ LEFT+51 should give as a result `left=1` if I'm not wrong. Actually I'd suggest not using LEFT but just use actual values like 100/110/.../150, I think this is actually clearer.
3. Remove the code for the test `'makes sure the hover image does not go off the left side of screen'`, I think it's useless with the test mentioned in step 2.

Ideally add a comment explaining all this :D So that next time we don't need to spend as much time thinking about that. Maybe this:
```
// LEFT is 100 and the screenshot's width is 300px.
// So hovering between 100 and 150px should give a result of left == 0.
// As soon as we pass 150px the left value should also move up.
```

(Now I'm looking at the test for TOP ahah)

Reply:
It's really awesome how you guys give such detailed comments. I really appreciate the effort behind this. I guess this is what keeps the contributors inspired. THANKS!
Let me try this again, I'll keep you posted.

Reply:
For a test for the TOP, I'd suggest just changing the existing test: https://github.com/firefox-devtools/profiler/blob/a16fb93e5c6913f812942f9f3af15c8356bc3284/src/test/components/TrackScreenshots.test.js#L90-L94

By just changing to:
```js
    expect(moveMouseAndGetTop(pageX)).toBe(0);
```

But then you can add another test to check a case where top is bigger than 0. To do this, as you probably understand now, we need to change what `getBoundingClientRect` returns. So this could possibly be something like this:
```js
  it('places the hover image in the center of the track if there is enough space', () => {
    const { setBoundingClientRectOffset, moveMouseAndGetTop } = setup();
    const containerTop = 100;
    const screenshotHeight = 150; // This value is defined in the fixtures.
    setBoundingClientRectOffset({ left: LEFT, top: containerTop });
    const pageX = LEFT;
    const expectedTop = containerTop + FULL_TRACK_SCREENSHOT_HEIGHT / 2 - screenshotHeight / 2;
    expect(moveMouseAndGetTop(pageX)).toBe(expectedTop);
  });
```
and in the `setup` function, implement this `setBoundingClientRectValues`:
```diff
diff --git a/src/test/components/TrackScreenshots.test.js b/src/test/components/TrackScreenshots.test.js
index 02a0755d..c38a620a 100644
--- a/src/test/components/TrackScreenshots.test.js
+++ b/src/test/components/TrackScreenshots.test.js
@@ -157,30 +157,44 @@ function setup(
 ) {
   const store = storeWithProfile(profile);
   const { getState, dispatch } = store;
   const flushRafCalls = mockRaf();
   const ctx = mockCanvasContext();
   jest
     .spyOn(HTMLCanvasElement.prototype, 'getContext')
     .mockImplementation(() => ctx);
+
+  let leftOffset = LEFT;
+  let topOffset = TOP;
   jest
     .spyOn(HTMLElement.prototype, 'getBoundingClientRect')
     .mockImplementation(() => {
       const rect = getBoundingBox(TRACK_WIDTH, FULL_TRACK_SCREENSHOT_HEIGHT);
       // Add some arbitrary X offset.
-      rect.left += LEFT;
-      rect.right += LEFT;
-      rect.x += LEFT;
-      rect.y += TOP;
-      rect.top += TOP;
-      rect.bottom += TOP;
+      rect.left += leftOffset;
+      rect.right += leftOffset;
+      rect.x += leftOffset;
+      rect.y += topOffset;
+      rect.top += topOffset;
+      rect.bottom += topOffset;
       return rect;
     });
 
+  function setBoundingClientRectOffset({
+    left,
+    top,
+  }: {
+    left: number,
+    top: number,
+  }) {
+    leftOffset = left;
+    topOffset = top;
+  }
+
   const renderResult = render(<Provider store={store}>{component}</Provider>);
   const { container } = renderResult;
 
   // WithSize uses requestAnimationFrame
   flushRafCalls();
 
   function screenshotHover() {
     return ensureExists(
@@ -219,16 +233,17 @@ function setup(
     getState,
     thread: profile.threads[0],
     store,
     screenshotHover,
     screenshotTrack,
     moveMouse,
     moveMouseAndGetLeft,
     moveMouseAndGetTop,
+    setBoundingClientRectOffset,
   };
 }
 
 /**
  * Take a thread full screenshot markers, and set some to "Unknown" in order to
  * create gaps in a screenshot track.
  */
 function _setScreenshotMarkersToUnknown(
```

Hope this makes sense!

Reply:
> It's really awesome how you guys give such detailed comments.

Ahah, to be honest this also forced me to understand better how this all works :-) In the end we get better tests, it's win-win ! Thanks again :)